### ISSUE TITLE
[RFC]: HunyuanImage 3.0 deployment

### ISSUE THREAD

**Issue by nussejzz** [vllm-project/vllm-omni#1483]
### Motivation.

HunyuanImage 3.0 is currently the state-of-the-art (SOTA) open-source model in the image generation and editing domain. It utilizes a massive 80B MoE backend via a Transfusion architecture. In `vllm-omni`, while the base DiT component has been merged (with AR pending) and Tensor Parallelism (TP) is supported, we face severe bottlenecks when scaling to production-grade disaggregated deployment:

(1) **Throughput and Efficiency under CFG**: The heavy computational burden of running an 80B MoE iteration loop (50 steps) requires significant optimization. The current standard batch duplication for Classifier-Free Guidance (CFG) causes severe compute waste due to uneven sequence padding between conditional and unconditional branches.
(2) **The Pipeline Parallelism (PP) Dilemma**: Storing and updating a full 80B MoE pushes VRAM boundaries. While the model authors desire a TP + PP + UP combination, applying PP to a DiT loop is highly problematic. It introduces extreme cyclic communication overhead because the latent tensor must be transferred from the last PP rank back to the first PP rank repetitively for 50 inference steps. A robust memory scaling alternative (e.g., Weight Offloading) must be evaluated and implemented.
(3) **High-Resolution Scaling**: The current implementation lacks support for Ulysses Parallelism (UP) and Ring Attention, which are strictly required for handling the massive token sequence lengths generated in high-resolution/4K image editing tasks.

For HunyuanImage 3.0, our goal is to solidify the DiT inference engine, finalize the cross-stage KV transfer, and introduce advanced parallelism (Packed-Attn, UP, Offloading) to maximize hardware utilization and achieve the speed needed for production viability.

### P0: Core Architecture & Foundation

- [ ] **Packed Attention CFG implementation (DiT Stage)** [Optimization]
  - Replace the naive `latent_model_input = torch.cat([latents] * cfg_factor)` duplication approach.
  - Implement a contiguous sequence packing approach (`packed-attn` / `cu_seqlens`), allowing Conditional and Unconditional streams to utilize varying subset lengths without adding zero-padding arithmetic waste into the MoE blocks.
- [ ] **KV Cache Transfer Integration for DiT** [Feature]
  - Finalize the integration of `OmniKVTransferManager` within the DiT pipeline to seamlessly receive and deserialize the massive Text Prefix context from the AR stage, ensuring zero re-computation of prompt/image conditions during the denoising loop.

### P1: Performance & Production Readiness

- [ ] **Pipeline Parallelism (PP) vs. Layerwise Weight Offloading** [Architecture]
  - Conduct a definitive overhead analysis to resolve the memory-wall issue. 
  - *Context:* We think PP communication overhead is excessively high due to the Stage N → Stage 0 cyclic transfer per timestep. 
  - *Action:* Prioritize and implement a highly optimized **Layerwise Weight Offloading** pipeline (using asynchronous prefetching streams from CPU RAM) as the primary alternative to PP for memory-constrained environments.
- [ ] **CFG-Parallel Implementation** [Feature]
  - Implement parallel dispatch logic specifically allowing DiT to calculate the Cond branch and Uncond branch synchronously across independently split GPUs, followed by a fast cross-node gathering reduction.

### P2: Advanced Scalability

- [ ] **Ulysses Sequence Parallelism (UP) Implementation** [Feature]
  - Implement Ulysses SP for the HunyuanImage 3.0 backend. By partitioning the spatial image tokens across GPUs, we can gracefully support massive resolution extensions without OOM.
- [ ] **Ring-Attention Integration** [Feature]
  - Implement Ring-Attention as a complementary feature to UP, further supporting extreme context windows required by complex text-to-image or image-to-image editing tasks.

### CC List.

@hsliuustc0106 @princepride @lishunyang12 @yuanheng-zhao 

### Any Other Things.

<img width="4839" height="8192" alt="Image" src="https://github.com/user-attachments/assets/f0214ba4-7999-411b-a392-1fe8c348d020" />

### Core difference between HunyuanImage 3.0 vs. Bagel

To contextualize the optimization focus, the fundamental structural differences between this model and our recent Bagel implementation must be highlighted:

1. **Native Transfusion (Single Prefix KV) vs. Frankenstein Contexts (Decoupled Trees)** 
   - *Bagel* acts as a decoupled router, utilizing 5 serial `forward` and `deepcopy` operations to compute varying contexts (Text-only, Text+Img, etc.), and managing them concurrently during the DiT phase.
   - *Hunyuan* is a native multi-modal model (Transfusion architecture). The Text, VAE, and ViT tokens are packed dynamically into a singular, immensely long context array using **Generalized Causal Attention** (where image patches can fully attend to each other simultaneously). Thus, Omni only needs to transfer and manage *one* massive Prefix KV, rather than multiple isolated feature trees.
2. **CFG Batched Splitting (Hunyuan) vs. `cu_seqlens` Muxing (Bagel)**
   - Because of its complex tree, *Bagel* computes CFG inside DiT by leveraging `cu_seqlens` and `packed-attn` natively, cleverly multiplexing shared context inside a monolithic KV structure to avoid batch duplication.
   - *Hunyuan* currently uses crude batch duplication (`torch.cat([latents] * cfg_factor)`) prior to the MoE. This doubles runtime VRAM activations and introduces vast padding waste. Bringing our Bagel-style `packed-attn` methodology into Hunyuan is our highest P0 priority for eliminating this specific bottleneck.
3. **MoE Dominance in DiT (13B Active vs. 7B Dense)**
   - *Bagel* runs diffusion on a relatively dense standard backend, meaning compute resources are fairly uniform per sequence length.
   - *Hunyuan's* 80B DiT loop heavily relies on MoE (activating ~13B per step). Without sophisticated Expert Parallelism (EP) intersecting with TP/UP inside the 50-loop timestep cycle, specific nodes will easily face Out-Of-Memory (OOM) deadlocks or massive compute bubbles.

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by princepride** [comment]
How about we draw a flowchart first? Furthermore, this model may also involve changes to the scheduler in the prompt rewriting part. Please also cc @ZJY0516 @natureofnature and hunyuan image at + dit contributor.

**Comment by nussejzz** [comment]
OK, I've compiled a preliminary version based on my understanding, and welcome any criticism and corrections.😊 @princepride

**Comment by hsliuustc0106** [comment]
from my point of view, I want to see when we can support batching/cudagraph/online/offline/disaggregated deployment/parallel strategy....

**Comment by nussejzz** [comment]
> from my point of view, I want to see when we can support batching/cudagraph/online/offline/disaggregated deployment/parallel strategy.... 

OK, I'll add them and try to solve as soon as possible with our teamates @hsliuustc0106

**Comment by princepride** [comment]
Yes, you can help Shuiquan implement ar part first.

**Comment by hsliuustc0106** [comment]
> Yes, you can help Shuiquan implement ar part first.

this is more urgent.

**Comment by nussejzz** [comment]
OK

**Comment by desenSunUBW** [comment]
Hello, do you consider using [MixFusion](https://dl.acm.org/doi/10.1145/3774934.3786420) to solve the resource waste problem for the mixed resolution serving scenario? For DiT model, it simply separates the sequence into multiple sub-sequences with identical length for all operators except self-attention. When calculating self-attention, it recovers the sub-sequences back to the original sequence and calculates the attention.

**Comment by hsliuustc0106** [comment]
Hi, guys, #759 merged but it seems the dit part has met some acc bugs #1489. Let's speed up the process for acc maintainance and performance improvement.

cc @princepride @nussejzz @Semmer2 @desenSunUBW
